File: examples/kotlin/projector-receipt/src/main/kotlin/dev/angzarr/examples/projector/ReceiptState.kt
package dev.angzarr.examples.projector

import examples.Domains.LineItem

/**
 * Immutable receipt projection state for Order domain.
 */
data class ReceiptState(
    val customerId: String = "",
    val items: List<LineItem> = emptyList(),
    val subtotalCents: Int = 0,
    val discountCents: Int = 0,
    val loyaltyPointsUsed: Int = 0,
    val finalTotalCents: Int = 0,
    val paymentMethod: String = "",
    val completedAt: com.google.protobuf.Timestamp? = null,
    val isCompleted: Boolean = false
) {
    companion object {
        fun empty(): ReceiptState = ReceiptState()
        const val POINTS_PER_DOLLAR = 10
    }
}

File: examples/typescript/projector-receipt/src/receipt-state.ts
/**
 * Receipt projection state.
 */
export interface LineItem {
  productId: string;
  name: string;
  quantity: number;
  unitPriceCents: number;
}

export interface ReceiptState {
  customerId: string;
  items: LineItem[];
  subtotalCents: number;
  discountCents: number;
  finalTotalCents: number;
  paymentMethod: string;
  loyaltyPointsEarned: number;
  completedAt: { seconds: number; nanos: number } | null;
  isCompleted: boolean;
}

export function emptyReceiptState(): ReceiptState {
  return {
    customerId: '',
    items: [],
    subtotalCents: 0,
    discountCents: 0,
    finalTotalCents: 0,
    paymentMethod: '',
    loyaltyPointsEarned: 0,
    completedAt: null,
    isCompleted: false,
  };
}
